compiler/next/test/resolution/testPoi.cpp

```diff
@@ -0,0 +1,604 @@
+ * limitations under the License.
+ */
+
+#include "chpl/parsing/parsing-queries.h"
+#include "chpl/resolution/resolution-queries.h"
+#include "chpl/resolution/scope-queries.h"
+#include "chpl/uast/Call.h"
+#include "chpl/uast/Comment.h"
+#include "chpl/uast/Identifier.h"
+#include "chpl/uast/Module.h"
+#include "chpl/uast/Variable.h"
+
+// always check assertions in this test
+#ifdef NDEBUG
+#undef NDEBUG
+#endif
+
+#include <cassert>
+
+using namespace chpl;
+using namespace parsing;
+using namespace resolution;
+using namespace uast;
+
+// test showing point-of-instantiation behavior
+static void test1() {
+  printf("test1\n");
+  Context ctx;
+  Context* context = &ctx;
+
+  auto path = UniqueString::build(context, "input.chpl");
+  std::string contents = "module M {\n"
+                         "  proc generic(param p) {\n"
+                         "    helper();\n"
+                         "  }\n"
+                         "}\n"
+                         "module N {\n"
+                         "  use M;\n"
+                         "  proc helper() { }\n"
+                         "  generic(1);\n"
+                         "}\n";
+  setFileText(context, path, contents);
+
+  const ModuleVec& vec = parse(context, path);
+  assert(vec.size() == 2);
+  const Module* m = vec[0]->toModule();
+  assert(m);
+  assert(m->numStmts() == 1);
+  const Module* n = vec[1]->toModule();
+  assert(n);
+  assert(n->numStmts() == 3);
+  const Function* mGeneric = m->stmt(0)->toFunction();
+  assert(mGeneric);
+  const Call* helperCall = mGeneric->stmt(0)->toCall();
+  assert(helperCall);
+  const Function* nHelper = n->stmt(1)->toFunction();
+  assert(nHelper);
+  const Call* genericCall = n->stmt(2)->toCall();
+  assert(genericCall);
+
+  // resolve module N
+  const ResolutionResultByPostorderID& rr = resolveModule(context, n->id());
+
+  // get the resolved function
+  const ResolvedFunction* rfunc =
+    resolveOnlyCandidate(context, rr.byAst(genericCall));
+  assert(rfunc);
+
+  const ResolvedFunction* rhelp =
+    resolveOnlyCandidate(context, rfunc->resolutionById.byAst(helperCall));
+  assert(rhelp);
+  assert(rhelp->functionId() == nHelper->id());
+}
+
+// testing the challenging program from issue 18081
+static void test2() {
+  printf("test2\n");
+  Context ctx;
+  Context* context = &ctx;
+
+  auto path = UniqueString::build(context, "input.chpl");
+  std::string contents = R""""(
+    module G {
+      proc foo(param arg:int, val:int) {
+        bar(arg);
+      }
+      proc foo(param arg:int, val:real) {
+        bar(arg);
+      }
+    }
+    module M1 {
+      use G;
+      proc bar(arg: int) { }
+      proc runM1() {
+        var x: int;
+        foo(1, x);
+      }
+    }
+
+    module M2 {
+      use G;
+      proc bar(arg: int) { }
+      proc runM2() {
+        var y: real;
+        foo(1, y);
+      }
+    }
+
+    module User {
+      use M1, M2;
+      proc main() {
+        runM1();
+        runM2();
```

Comment:
Was the original bug that we'd end up calling the wrong foo (that is the int variant) or was it that we'd end up calling the right foo (the real variant) but that would end up calling the wrong bar?

Reply:
It's discussed in #18081 -- the bug is that runM2 calls M1.bar rather than M2.bar. It's a bit of a terminology issue to ask if it's calling the right foo since it calls the wrong instantiation of foo. But it is calling a foo with the right signature (real vs int second argument).